### ISSUE TITLE
【Task.31】DPO / Reward Modeling - RFC

### ISSUE THREAD

**Issue by shanyulu** [redai-studio/Relax#199]
# 【Task.31】DPO / Reward Modeling - RFC

| 项 | 固定值 |
| --- | --- |
| 状态 | Draft；未开始实现、未提交代码 |
| 认领人 | shanyulu
| 仓库 / 审计基线 | `redai-infra/Relax` `main@0bc99af8dd39de8fd99c588a98b3f3a463bc818c` |
| 拟用分支 | `feat/task31-dpo-rm-shanyulu` |
| 运行方式 | 独立 host venv；不使用容器 |
| 交付 | RFC → 分阶段 PR → recipe、数据说明、曲线、评测报告 |

本文中的新参数、类型和脚本均为提案，不表示当前仓库已经实现。

## 1. 决策与范围

在 SFT 离线训练链路上增加 DPO 与 Reward Model（RM），复用 producer、TransferQueue、Megatron 动态批处理、checkpoint 和训练循环；只新增 pair dataset/batch adapter、loss、离线 RM 分支、参数和 recipe。

- DPO：标准带 reference 的 DPO；reference-free 仅作为 reference log-ratio 为零的独立消融，不等同 SimPO/ORPO。
- RM：对同一 preference pair 训练 scalar reward，使 held-out 上 `score(chosen) > score(rejected)`。

本期不做在线采样、PPO/GAE、Rollout/SGLang、权重同步、live 双模型 reference、多模态 preference、CP>1、margin/pointwise/multi-head RM，亦不以 7B 以上双模型训练作为验收。初版对上述配置 fail-fast，不允许静默退化。

## 2. 现有链路与角色

```text
PreferenceDataset -> SFT producer -> TransferQueue -> Megatron data.py
                  -> actor / critic forward -> loss.py -> checkpoint
```

`SFTStreamingDataset`/`ProcessedSample` 当前只表示单序列；`loss_function` 是统一损失入口。RM 复用 `get_model_provider_func(role="critic")` 的 `LinearForLastLayer(hidden, 1)`，但不能走 PPO `train_critic` 的 value/GAE 路径。

| 目标 | 服务组合 | 不启动 |
| --- | --- | --- |
| DPO | `{sft, actor}` | rollout、reference、advantages |
| RM | `{sft, critic}` | rollout、actor、reference、advantages |

`sft` 保留为内部 producer 角色名以兼容已有资源和生命周期；训练目标由 `loss_type` 区分。RM 新增离线 critic 直训分支，直接计算 pairwise loss。

## 3. Pair 数据与调度契约

V1 输入限制：chosen/rejected 的非末尾消息必须逐项相同，且仅最后一条可学习 assistant completion 不同；不满足即拒绝。渲染复用 SFT 的 `render_with_loss_mask`，以 assistant loss-mask 定义 response，不依赖 `tokenize(prompt)` 必为完整序列前缀的假设。

```text
TQ index = 1 个 PreferencePair
global_batch_size / micro_batch_size = pair 数，不是 sequence 数
pair_cost = chosen_total_length + rejected_total_length
DP 与动态 microbatch 按 pair_cost 分配；展开发生在分配之后
```

```text
PreferencePairBatch(P pairs)
  pair_ids:                 [P]
  chosen_tokens/rejected_tokens: [P] / [P]
  chosen_masks/rejected_masks:   [P] / [P]
  pair_costs:               [P]
  ref_logp_sums:            [P, 2]                 # DPO/reference only

data.py::get_preference_batch
  -> [chosen_0, rejected_0, ..., chosen_(P-1), rejected_(P-1)]
  -> tokens/loss_masks/total_lengths/response_lengths: [2P]
  -> pair_sequence_offsets: [[0,1], [2,3], ...]
```

`get_preference_batch` 在当前 `get_batch` 完成 TQ 读取且 DP/microbatch 已选定后、CP/THD packing 前展开。THD flatten 后沿用每 sequence 的 span；logprob helper 返回各 sequence response logprob 后，按 `pair_sequence_offsets` 还原 chosen/rejected 的 sequence sum。chosen/rejected 永不串成一条因果序列，彼此不可见。

截断顺序固定：共享左截 prompt → 两支各自右截 completion → `max_seq_length`。label 采用现有 next-token 左移语义，末位不计 loss；任一支没有有效 response token 则整对丢弃。日志记录 seen/kept/empty/oversize、长度分位数与截断比例。

## 4. Reference artifact

带 reference 的 DPO 预计算产物为逐 pair、逐 branch 记录；全局 manifest 和记录哈希同时校验。

| 字段 | 含义 |
| --- | --- |
| `pair_id` | 稳定来源 ID |
| `input_ids_sha256` / `response_mask_sha256` | 每支实际 token 与 mask 的字节哈希 |
| `response_token_count` | 参与 logprob 的 token 数 |
| `ref_logp_sum` / `ref_logp_dtype` | response logprob 总和；FP32 累积 |
| `reference_checkpoint_revision` | 冻结 reference 权重 |
| `tokenizer_revision` / `chat_template_sha256` | tokenization 契约 |

预计算固定 `model.eval()`、dropout disabled、temperature=1.0、无 sampling、仅 response token、FP32 `log_softmax` 累加。训练前重新计算 input/mask hash；任一项不匹配即拒绝。现有 response-aligned logprob helper 优先复用，禁止另写一套 logits/token 切片。

## 5. 损失、梯度与模型头

```text
z_i     = beta * ((logπc - logπr) - (logπref,c - logπref,r))
L_dpo,i = -logsigmoid(z_i)
L_rm,i  = -logsigmoid(score(chosen) - score(rejected))
```

reference-free DPO 将 reference 括号置零。每支 logprob 仅在 response mask 上求和；RM score 取最后一个有效 response token，不能取 prompt、padding 或固定末位。

在 DDP 平均梯度契约下，设本 rank 的 pair loss sum 为 `S_r`、pair count 为 `n_r`、`N=all_reduce(sum(n_r))`、DP world size 为 `D`：

```text
backward_loss = S_r * D / N
reported_loss = all_reduce(sum(S_r)) / N
```

因此不使用 `local_loss.mean()` 或 rank mean 的平均。初版 `CP=1`；必须以 uneven-DP pair count 的单进程完整 batch 为参考，验证分布式参数梯度和 reported loss 对齐。

审计基线中 `LinearForLastLayer` 的实际实现为 `normal_(mean=0.0, std=0.02)` / zero bias；seed 固定 42，并记录初始化后 head weight 的 mean/std 与 bias。base model 全参数训练。RM checkpoint 必须保存 base、head、optimizer、`training_kind=rm` 与 head schema；resume 校验这些元数据，DPO actor checkpoint 不可加载为 RM。增加 save→reload 后 held-out score `allclose` 测试。

## 6. 改动与 PR 拆分

| PR | 改动 | 验收 |
| --- | --- | --- |
| 1 | `PreferencePair` dataset、渲染/截断、TQ pair contract、batch expansion、CPU tests | pair 不拆分；mask/label/span 正确 |
| 2 | DPO 参数、artifact 校验、loss、uneven-DP reducer、recipe | 独立 PyTorch 的逐元素和梯度对齐 |
| 3 | 离线 RM critic 分支、pairwise loss、checkpoint/eval、recipe | 不进入 PPO/GAE；save/reload 与 held-out 指标通过 |

预计涉及：`relax/engine/sft/`、`relax/components/sft.py`、`relax/backends/megatron/{data.py,loss.py,actor.py,model_provider.py}`、`relax/{core/registry.py,utils/arguments.py}`、`scripts/training/preference/` 与对应 tests。现有 SFT 脚本及其 focused suite 必须通过。

## 7. 固定数据、环境与实验协议

数据：[HuggingFaceH4/ultrafeedback_binarized](https://huggingface.co/datasets/HuggingFaceH4/ultrafeedback_binarized) `3949bf5f8c17c394422ccfab0c31ea9c20bdeb85`（MIT），`train_prefs` 61,135 对、`test_prefs` 2,000 对；模型：[Qwen/Qwen3-0.6B](https://huggingface.co/Qwen/Qwen3-0.6B) `c1899de289a04d12100db370d81485cdf75e47ca`（Apache-2.0）。不提交模型或数据副本。

机器固定为 2× NVIDIA A800 80GB PCIe（driver 580.82.07，跨 NUMA `SYS`、无 NVLink），2× Xeon Gold 6348、112 logical CPUs、约 1 TiB RAM。主训练和 uneven-DP parity 都在该双卡环境完成，不以单卡日志替代分布式验收。

运行使用独立 host venv：Python 3.12.3、PyTorch 2.11.0/CUDA 13.0、NCCL 2.28.9、Ray 2.56.1、SGLang 0.5.12.post1、Megatron-Core `0.18.0+85bced0ae`、Megatron-Bridge `0.5.0+2faedbf6`。每次实验必须生成 `env/`：`pip freeze --all`、`pip inspect`、Python/CUDA/NCCL 版本、Megatron-Bridge checkout SHA、Relax patch tree 的 `git diff --binary` SHA256、安装 wheel 的 SHA256 与完整安装命令。环境通过以下锁定入口重建；锁文件和 wheel hash 随 recipe 交付：

```bash
python3.12 -m venv task31_env
task31_env/bin/python -m pip install --require-hashes -r scripts/training/preference/requirements-task31.lock
```

共同约束：10,000 train pairs、2,000 held-out pairs、BF16、`TP=1/DP=2/PP=1/CP=1`、`max_prompt_length=1024`、`max_response_length=1024`、`max_seq_length=2048`、`max_tokens_per_gpu=8192`、pair global batch 16、micro batch 1/GPU、seed 42、AdamW、beta=(0.9,0.95)、clip-grad=1.0、warmup=3%。

| 配方 | 训练参数 |
| --- | --- |
| DPO | `beta=0.1`；initial policy = reference checkpoint；full-parameter；lr `1e-6`；weight decay `0`；625 optimizer steps（1 epoch）；eval/save every 125 steps |
| RM | full-parameter + 新 head；lr `1e-5`；weight decay `0.01`；625 optimizer steps（1 epoch）；eval/save every 125 steps |

拟议入口（实现后才可运行）：

```bash
bash scripts/training/preference/prepare_ultrafeedback_pairs.sh \
  --dataset HuggingFaceH4/ultrafeedback_binarized \
  --revision 3949bf5f8c17c394422ccfab0c31ea9c20bdeb85 \
  --model Qwen/Qwen3-0.6B \
  --model-revision c1899de289a04d12100db370d81485cdf75e47ca \
  --train-pairs 10000 --max-prompt-length 1024 --max-response-length 1024 --max-seq-length 2048

CUDA_VISIBLE_DEVICES=0,1 bash scripts/training/preference/run-qwen3-0.6b-dpo-2xgpu.sh
CUDA_VISIBLE_DEVICES=0,1 bash scripts/training/preference/run-qwen3-0.6b-rm-2xgpu.sh
```

脚本将保存完整 `ray job submit` 命令、commit、seed、拓扑、版本和 artifact manifest；RFC 不把拟议命令写成已运行结果。

## 8. 验证与失败条件

- CPU：message-prefix 限制、渲染/mask/label、截断、空 completion、artifact hash、pair 原子性；DPO loss/gradient 与独立 PyTorch 实现逐元素对齐；RM score 选择与 padding 不变性。
- DP：2 GPU、uneven pair count、长度不等 pair，至少 10 optimizer steps；与单进程完整 batch 做 loss/gradient parity。
- RM：报告训练前/后 pairwise accuracy、`mean(chosen-rejected)`、95% bootstrap CI、loss、有效/丢弃 pair 数；验收为 mean margin `>0` 且 accuracy 95% CI 下界 `>0.5`。
- DPO：报告 loss、implicit reward margin、chosen/rejected response logp、pairwise accuracy 与完整命令；reference-free 单列。
- 回归：相关 SFT focused tests、pre-commit 与新增测试全部通过；RM checkpoint save→reload score 对齐。

schema/artifact 不匹配、pair 跨 rank、无有效 completion、NaN/Inf、TQ 残留 partition/超时或 checkpoint schema 不一致均为失败，不得跳过后报成功。

## 9. 参考

- [官方任务说明](https://github.com/redai-infra/community/blob/main/contributor-program/2026-cohort-1/official-task.md)
- [Relax SFT producer](https://github.com/redai-infra/Relax/blob/0bc99af8dd39de8fd99c588a98b3f3a463bc818c/relax/components/sft.py)
- [Relax Megatron loss dispatch](https://github.com/redai-infra/Relax/blob/0bc99af8dd39de8fd99c588a98b3f3a463bc818c/relax/backends/megatron/loss.py)
- [DPO](https://arxiv.org/abs/2305.18290)

**Comment by SigureMo** [comment]
#208 入选，本 RFC close
